src/components/ui/TableOfContent.tsx

```diff
@@ -13,28 +13,65 @@ export type TableOfContentProps = {
 
 export const TableOfContent = ({ headers }: TableOfContentProps) => {
   const router = useRouter();
+  const onThisPageID = "on-this-page-heading";
   const [currentHash, setCurrentHash] = useState<Nullable<string>>(null);
 
+  // useEffect(() => {
+  //   const handleHashChange = (event: HashChangeEvent) => {
+  //     setCurrentHash(event.newURL.split("#")[1]);
+  //   };
+
+  //   window.addEventListener("hashchange", handleHashChange);
+  //   return () => {
+  //     window.removeEventListener("hashchange", handleHashChange);
+  //   };
+  // }, []);
+
+  // useEffect(() => {
+  //   if (!router.isReady) return;
+
+  //   setCurrentHash(router.asPath.split("#")[1]);
+  // }, [router]);
+
   useEffect(() => {
-    const handleHashChange = (event: HashChangeEvent) => {
-      setCurrentHash(event.newURL.split("#")[1]);
+    const setCurrent: IntersectionObserverCallback = (entries) => {
+      for (const entry of entries) {
+        if (entry.isIntersecting) {
+          const { id } = entry.target;
+          if (id === onThisPageID) continue;
+          setCurrentHash(entry.target.id);
+          break;
+        }
+      }
     };
 
-    window.addEventListener("hashchange", handleHashChange);
-    return () => {
-      window.removeEventListener("hashchange", handleHashChange);
+    const observerOptions: IntersectionObserverInit = {
+      // Negative top margin accounts for `scroll-margin`.
+      // Negative bottom margin means heading needs to be towards top of viewport to trigger intersection.
+      rootMargin: "-100px 0% -66%",
+      threshold: 1,
     };
-  }, []);
 
-  useEffect(() => {
-    if (!router.isReady) return;
+    const headingsObserver = new IntersectionObserver(
+      setCurrent,
+      observerOptions
+    );
+
+    // Observe all the headings in the main page content.
+    document
+      .querySelectorAll("article :is(h1,h2,h3)")
```

Comment:
@iamnamananand996 We didn't limit the usage of header, so the header may be h5 at maximum, we have to count this situation into consideration

```suggestion
      .querySelectorAll("article :is(h1,h2,h3,h4,h5)")
```